### PR TITLE
update_kubernetes_version: standalone script v2

### LINKED ISSUE #4392: Write a script to automatically propose PR's due to a new Kubernetes release
I would love a script that we can run from a cronjob or Jenkins that monitors https://github.com/kubernetes/kubernetes/releases for new releases.

The script should propose a PR for any release newer than NewestKubernetesVersion:
 https://github.com/kubernetes/minikube/blob/f8fa0748c5a474bc2a62816ce318a6740fc00fea/pkg/minikube/constants/constants.go#L168 

If the release is a non-beta release, it should also change DefaultKubernetesVersion in the same PR:

https://github.com/kubernetes/minikube/blob/f8fa0748c5a474bc2a62816ce318a6740fc00fea/pkg/minikube/constants/constants.go#L165

This script should either build upon or replace  https://github.com/kubernetes/minikube/blob/master/hack/kubernetes_version/update_kubernetes_version.go - which makes the changes necessary in our code base, but is designed for interactive use and does not know how to look for new Kubernetes releases.

We can provide a github token via an environment variable.

### PR DESCRIPTION
**Pull request by prezha**
fixes #4392 (iteration 1 & 2)

### notes:
- `UPDATE_TARGET` env var controlles what will be updated - options:
  = unset/absent, `fs`, or `all`: local filesystem repo will be updated - similarly to the old behaviour
  = `gh` or `all`: a pr against origin (`kubernetes/minikube`) github repo would be automatically created, but only if the same one does not exist already, and will also 'mention' the issue #4392
- `GITHUB_TOKEN` is required for `gh` or `all` `UPDATE_TARGET`

- desired changes are defined as `plan`, which is flexible (ie, supports regex and go text templates) and placed at the top of the script - easy to amend and extend; parseability checks in place

- aob:
  = used `github.com/google/go-github/v32/github`: pure go code ;) - no need to call `git` or `gh` cli from the script
  = used `k8s.io/klog/v2` instead of `glog`
  = `go.mod` & `go.sum` amended accordingly (for klog/v2 and github/v32)
  = cron job script could contain something like: `rm -rf ~/minikube && git clone https://github.com/kubernetes/minikube.git ~/minikube && cd ~/minikube/hack/kubernetes_version && go run update_kubernetes_version.go`

- all `make test` passed

- testing & example output:
### 1st run:

```
I1003 00:52:44.696940    1777 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 00:52:44.697055    1777 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 00:52:44.697581    1777 update_kubernetes_version.go:172] Local repo updated
```

### git diff:

```
diff --git a/pkg/minikube/constants/constants.go b/pkg/minikube/constants/constants.go
index fd330a96d..a533e0e09 100644
--- a/pkg/minikube/constants/constants.go
+++ b/pkg/minikube/constants/constants.go
@@ -27,10 +27,10 @@ import (

 const (
        // DefaultKubernetesVersion is the default Kubernetes version
-       DefaultKubernetesVersion = "v1.19.2"
+       DefaultKubernetesVersion = "v1.23.456"
        // NewestKubernetesVersion is the newest Kubernetes version to test against
        // NOTE: You may need to update coreDNS & etcd versions in pkg/minikube/bootstrapper/images/images.go
-       NewestKubernetesVersion = "v1.19.2"
+       NewestKubernetesVersion = "v2.34.457-rc.89"
        // OldestKubernetesVersion is the oldest Kubernetes version to test against
        OldestKubernetesVersion = "v1.13.0"
        // DefaultClusterName is the default nane for the k8s cluster
diff --git a/site/content/en/docs/commands/start.md b/site/content/en/docs/commands/start.md
index 87538fb59..b1f036128 100644
--- a/site/content/en/docs/commands/start.md
+++ b/site/content/en/docs/commands/start.md
@@ -67,7 +67,7 @@ minikube start [flags]
       --interactive                       Allow user prompts for more information (default true)
       --iso-url strings                   Locations to fetch the minikube ISO from. (default [https://storage.googleapis.com/minikube/iso/minikube-v1.13.1.iso,https://github.com/kubernetes/minikube/releases/download/v1.13.1/minikube-v1.13.1.iso,https://kubernetes.oss-cn-hangzhou.aliyuncs.com/minikube/iso/minikube-v1.13.1.iso])
       --keep-context                      This will keep the existing kubectl context and will create a minikube context.
-      --kubernetes-version string         The Kubernetes version that the minikube VM will use (ex: v1.2.3, 'stable' for v1.19.2, 'latest' for v1.19.2). Defaults to 'stable'.
+      --kubernetes-version string         The Kubernetes version that the minikube VM will use (ex: v1.2.3, 'stable' for v1.23.456, 'latest' for v2.34.457-rc.89). Defaults to 'stable'.
       --kvm-gpu                           Enable experimental NVIDIA GPU support in minikube
       --kvm-hidden                        Hide the hypervisor signature from the guest in minikube (kvm2 driver only)
       --kvm-network string                The KVM network name. (kvm2 driver only) (default "default")

```

### 2nd run:

```
I1003 00:53:30.207909    3317 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 00:53:30.208360    3317 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 00:53:30.209539    3317 update_kubernetes_version.go:170] Local repo update skipped: nothing changed
```

### 1st run:

```
I1003 01:06:26.158314   21354 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 01:06:26.158472   21354 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 01:06:34.204164   21354 update_kubernetes_version.go:289] PR commit '0d49507f0f4d6fb8c6f8cf646a65a737371da2e8' created: https://github.com/prezha/minikube/commit/0d49507f0f4d6fb8c6f8cf646a65a737371da2e8
I1003 01:06:34.921082   21354 update_kubernetes_version.go:303] PR branch 'update-kubernetes-version_0d49507' created: https://api.github.com/repos/prezha/minikube/git/refs/heads/update-kubernetes-version_0d49507
I1003 01:06:36.615138   21354 update_kubernetes_version.go:199] PR created: https://github.com/triplepoint-tech/minikube/pull/3
```

### 2nd run:

```
I1003 01:06:52.070881   22035 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 01:06:52.071322   22035 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 01:06:52.434969   22035 update_kubernetes_version.go:190] PR create skipped: already exists (https://github.com/triplepoint-tech/minikube/pull/3)
```

### CODE REVIEW

File: hack/kubernetes_version/update_kubernetes_version.go [hunk 1]
```diff
@@ -14,159 +14,442 @@ See the License for the specific language governing permissions and
...
 	"path/filepath"
 	"regexp"
 	"strings"
+	"text/template"
 	"time"
 
-	"github.com/golang/glog"
+	"golang.org/x/oauth2"
+
+	"k8s.io/klog/v2"
```
Comment (medyagh) [inline hack/kubernetes_version/update_kubernetes_version.go:45]: I recommend keep using glog till we deprecate glog (that is a WIP PR it is better to let that PR handle migration from glog to klog)

**Review by medyagh — CHANGES_REQUESTED** [review]

File: hack/kubernetes_version/update_kubernetes_version.go (hunk 1 above)
Comment (prezha) [reply hack/kubernetes_version/update_kubernetes_version.go]: ok, thanks @medyagh for your review and comment - i've switched back from using klog to glog

**Review by medyagh — COMMENTED** [review]
thank you

**Comment by prezha** [comment]
@medyagh you are most welcome, thank you for your review!
please let me know if there are any issues running this script (scheduled)

**Pull request merged**